File: apps/example/android/app/src/main/java/com/bridgekit/example/bridgekit/BridgekitDemoInitializer.kt
package com.bridgekit.example.bridgekit

import android.util.Log
import com.bridgekit.example.generated.BridgekitDemoFeatureContract
import com.bridgekit.example.generated.BridgekitDemoHostContract
import com.bridgekit.example.generated.BridgekitDemoReverseContract
import com.bridgekit.example.generated.GetGreetingParams
import com.bridgekit.example.generated.GreetFromJsParams
import com.bridgekit.example.generated.OnNativeEventParams
import io.github.malopezr7.bridgekit.core.BridgeKit
import io.github.malopezr7.bridgekit.core.Scope
import io.github.malopezr7.bridgekit.discovery.BridgeKitHost
import io.github.malopezr7.bridgekit.runtime.BridgeValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.take
import kotlinx.coroutines.launch

private const val TAG = "BridgeKitDemo"
private const val TAG_REVERSE = "BridgeKitReverse"
private val appScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

/**
 * Initializes BridgeKit for the example app.
 * Called from MainApplication.onCreate BEFORE React initializes.
 */
object BridgekitDemoInitializer {

    fun init(host: BridgeKitHost) {
        val bridgeKit = BridgeKit.default
        // Run ServiceLoader discovery (no-op if no META-INF/services present)
        try {
            bridgeKit.initialize(host)
        } catch (e: Exception) {
            Log.w(TAG, "ServiceLoader discovery failed (expected in bare app): ${e.message}")
        }

        // Provide the native demo-host contract at Global scope
        bridgeKit.provide(
            definition = BridgekitDemoHostContract,
            scope = Scope.Global,
            eager = true,
        ) {
            DemoHostImpl()
        }
        Log.i(TAG, "BridgeKit: DemoHost provided at Global scope")

        // After a delay, consume the JS-provided demo-feature contract and call getGreeting.
        // Validates the JS-provides / native-consumes Async direction (cross-direction scenario).
        appScope.launch {
            delay(15_000)
            try {
                val feature = bridgeKit.consume(BridgekitDemoFeatureContract, Scope.Global)
                val result = feature.getGreeting(GetGreetingParams(name = "BridgeKit"))
                Log.i(TAG, "JS said: $result")
            } catch (e: Exception) {
                Log.e(TAG, "native→JS consume failed: ${e.message}", e)
            }
        }

        // Consume the JS-provided demo-reverse contract.
        // Exercises ALL four marker directions (JS provides, native consumes):
        //   Async  — greetFromJs
        //   Void   — onNativeEvent
        //   Stream — jsCounter
        //   State  — jsStatus
        appScope.launch {
            try {
                val reverse = bridgeKit.consume(BridgekitDemoReverseContract, Scope.Global)

                // ---- Async (doubles as the readiness probe) ----
                // The RN bundle can take a while to load+execute on the emulator,
                // so retry until the JS dispatcher has provided the contract (up to ~90 s).
                val deadline = System.currentTimeMillis() + 90_000
                var asyncOk = false
                while (!asyncOk && System.currentTimeMillis() < deadline) {
                    try {
                        val greeting = reverse.greetFromJs(GreetFromJsParams(name = "Android"))
                        Log.i(TAG_REVERSE, "async=$greeting")
                        asyncOk = true
                    } catch (e: Exception) {
                        delay(3_000)
                    }
                }
                if (!asyncOk) Log.e(TAG_REVERSE, "async never became ready within 90s")

                // ---- Void (fire-and-forget from native's perspective) ----
                try {
                    reverse.onNativeEvent(OnNativeEventParams(type = "native-tap", payload = "button_a"))
                    Log.i(TAG_REVERSE, "void fired type=native-tap")
                } catch (e: Exception) {
                    Log.e(TAG_REVERSE, "void fire failed: ${e.message}", e)
                }

                // ---- Stream: collect first 5 ticks from the JS counter ----
                reverse.jsCounter()
                    .take(5)
                    .onEach { v -> Log.i(TAG_REVERSE, "stream tick=$v") }
                    .catch { e -> Log.e(TAG_REVERSE, "stream failed: ${e.message}") }
                    .launchIn(appScope)

                // ---- State: observe jsStatus until cancelled ----
                reverse.jsStatus
                    .catch { e -> Log.e(TAG_REVERSE, "state failed: ${e.message}") }
                    .onEach { bv ->
                        when (bv) {
                            is BridgeValue.Available ->
                                Log.i(TAG_REVERSE, "state=${bv.value}")
                            is BridgeValue.Initial ->
                                Log.i(TAG_REVERSE, "state=initial(${bv.value})")
                            is BridgeValue.Replacing ->
                                Log.i(TAG_REVERSE, "state=replacing(${bv.lastKnown})")
                            is BridgeValue.Unprovided ->
                                Log.i(TAG_REVERSE, "state=unprovided(${bv.lastKnown})")
                        }
                    }
                    .launchIn(appScope)

            } catch (e: Exception) {
                Log.e(TAG_REVERSE, "reverse consume failed: ${e.message}", e)
            }
        }

        Log.i(TAG, "BridgeKit initialized. dump=${bridgeKit.dump()}")
    }
}

File: apps/example/ios/BridgeKitExampleDemo/Sources/BridgekitDemoInitializer.swift
import BridgeKit
import Foundation

private let bridgeKitDemoTag = "BridgeKitDemo"
private let bridgeKitReverseTag = "BridgeKitReverse"
private let bridgeKitJsInfoTag = "BridgeKitJsInfo"

enum BridgekitDemoInitializer {
  static func configure() {
    let bridgeKit = BridgeKitRuntime.default

    _ = bridgeKit.provide(BridgekitDemoHostContract(), scope: .global) {
      DemoHostImpl()
    }
    print("[\(bridgeKitDemoTag)] DemoHost provided at global scope")

    Task {
      try? await Task.sleep(nanoseconds: 15_000_000_000)
      do {
        let feature = bridgeKit.consume(BridgekitDemoFeatureContract(), scope: .global)
        let result = try await feature.getGreeting(GetGreetingParams(name: "BridgeKit"))
        print("[\(bridgeKitDemoTag)] JS greeting=\(result)")
      } catch {
        print("[\(bridgeKitDemoTag)] native-to-JS feature consume failed: \(error)")
      }
    }

    Task {
      do {
        let reverse = bridgeKit.consume(BridgekitDemoReverseContract(), scope: .global)
        let deadline = Date().addingTimeInterval(90)
        var asyncOk = false

        while !asyncOk && Date() < deadline {
          do {
            let greeting = try await reverse.greetFromJs(GreetFromJsParams(name: "iOS"))
            print("[\(bridgeKitReverseTag)] async=\(greeting)")
            asyncOk = true
          } catch {
            try? await Task.sleep(nanoseconds: 3_000_000_000)
          }
        }

        if !asyncOk {
          print("[\(bridgeKitReverseTag)] async never became ready within 90s")
        }

        reverse.onNativeEvent(OnNativeEventParams(type: "native-tap", payload: "button_a"))
        print("[\(bridgeKitReverseTag)] void fired type=native-tap")

        Task {
          var count = 0
          for await value in reverse.jsCounter() {
            print("[\(bridgeKitReverseTag)] stream tick=\(value)")
            count += 1
            if count >= 5 { break }
          }
        }

        Task {
          for await value in reverse.jsStatus {
            switch value {
            case .available(let status):
              print("[\(bridgeKitReverseTag)] state=\(status)")
            case .initial(let status):
              print("[\(bridgeKitReverseTag)] state=initial(\(status))")
            case .replacing(let last):
              print("[\(bridgeKitReverseTag)] state=replacing(\(String(describing: last)))")
            case .unprovided(let last):
              print("[\(bridgeKitReverseTag)] state=unprovided(\(String(describing: last)))")
            }
          }
        }
      } catch {
        print("[\(bridgeKitReverseTag)] reverse consume failed: \(error)")
      }
    }

    Task {
      let jsInfo = bridgeKit.consume(BridgekitDemoJsinfoContract(), scope: .global)
      let deadline = Date().addingTimeInterval(90)
      var ready = false

      while !ready && Date() < deadline {
        do {
          let version = try await jsInfo.getReactNativeVersion()
          let level = try await jsInfo.getUserLevel()
          let segments = try await jsInfo.getUserSegments()
          print(
            "[\(bridgeKitJsInfoTag)] rnVersion=\(version) userLevel=L\(Int(level.level)) \(level.label) segments=\(segments.joined(separator: ","))"
          )
          ready = true
        } catch {
          try? await Task.sleep(nanoseconds: 3_000_000_000)
        }
      }

      if !ready {
        print("[\(bridgeKitJsInfoTag)] getReactNativeVersion never became ready within 90s")
      }

      Task {
        var count = 0
        for await value in jsInfo.clockTicks() {
          print("[\(bridgeKitJsInfoTag)] clock tick=\(value)")
          count += 1
          if count >= 5 { break }
        }
      }
    }

    print("[\(bridgeKitDemoTag)] BridgeKit initialized. dump=\(bridgeKit.dump())")
  }
}

private final class DemoHostImpl: BridgekitDemoHost {
  private var counterValue: Double = 0
  private let counterLock = NSLock()
  private var counterContinuations: [UUID: AsyncStream<Double>.Continuation] = [:]
  private let counterContinuationLock = NSLock()

  private var echoContinuations: [UUID: AsyncStream<String>.Continuation] = [:]
  private let echoLock = NSLock()

  func ping(_ params: PingParams) async throws -> PingResult {
    PingResult(
      reply: "pong: \(params.message)",
      epoch: Double(Date().timeIntervalSince1970 * 1000)
    )
  }

  func increment() async throws -> Double {
    counterLock.lock()
    counterValue += 1
    let next = counterValue
    counterLock.unlock()

    pushCounter(next)
    return next
  }

  func say(_ params: SayParams) {
    let echoed = params.text.uppercased()
    print("[\(bridgeKitDemoTag)] DemoHostImpl.say: \(params.text) -> \(echoed)")
    echoText(echoed)
  }

  func ticker() -> AsyncStream<Double> {
    AsyncStream { continuation in
      let task = Task {
        var tick: Double = 0
        while !Task.isCancelled {
          try? await Task.sleep(nanoseconds: 1_000_000_000)
          tick += 1
          continuation.yield(tick)
        }
      }
      continuation.onTermination = { _ in task.cancel() }
    }
  }

  func echoes() -> AsyncStream<String> {
    AsyncStream { continuation in
      let id = UUID()
      echoLock.lock()
      echoContinuations[id] = continuation
      echoLock.unlock()

      continuation.onTermination = { [weak self] _ in
        guard let self else { return }
        self.echoLock.lock()
        self.echoContinuations.removeValue(forKey: id)
        self.echoLock.unlock()
      }
    }
  }

  var counter: AsyncStream<Double> {
    AsyncStream { continuation in
      let id = UUID()

      counterLock.lock()
      let current = counterValue
      counterLock.unlock()
      continuation.yield(current)

      counterContinuationLock.lock()
      counterContinuations[id] = continuation
      counterContinuationLock.unlock()

      continuation.onTermination = { [weak self] _ in
        guard let self else { return }
        self.counterContinuationLock.lock()
        self.counterContinuations.removeValue(forKey: id)
        self.counterContinuationLock.unlock()
      }
    }
  }

  private func pushCounter(_ value: Double) {
    counterContinuationLock.lock()
    let continuations = Array(counterContinuations.values)
    counterContinuationLock.unlock()

    for continuation in continuations {
      continuation.yield(value)
    }
  }

  private func echoText(_ text: String) {
    echoLock.lock()
    let continuations = Array(echoContinuations.values)
    echoLock.unlock()

    for continuation in continuations {
      continuation.yield(text)
    }
  }
}
